File: js_port/fs_js/src/block/ListExpression.js
const { ExpressionBlock } = require('./ExpressionBlock');
const { ArrayFsList } = require('../model/FsList');
const { ensureTyped, makeValue } = require('../core/value');
const { FSDataType } = require('../core/fstypes');

class ListExpression extends ExpressionBlock {
  constructor() {
    super();
    this.ValueExpressions = [];
  }

  evaluate(provider) {
    const values = this.ValueExpressions.map((expr) => ensureTyped(expr.evaluate(provider)));
    const list = new ArrayFsList(values);
    return makeValue(FSDataType.List, list);
  }

  getChilds() {
    return this.ValueExpressions.slice();
  }

  asExpressionString(provider) {
    const parts = this.ValueExpressions.map((expr) => expr.asExpressionString(provider));
    return `[${parts.join(', ')}]`;
  }
}

module.exports = {
  ListExpression
};

File: FunscScript/Block/ListExpression.cs
using FunscScript.Core;
using FunscScript.Model;
using System.Text;
using Newtonsoft.Json.Serialization;

namespace FunscScript.Block
{
    public class ListExpression:ExpressionBlock
    {
       
        public ExpressionBlock[] ValueExpressions;

       
        public override object Evaluate(IFsDataProvider provider )
        {
            var lst = ValueExpressions.Select(x => x.Evaluate(provider)).ToArray();
            return new ArrayFsList(lst);
        }
        public override IList<ExpressionBlock> GetChilds()
        {
            var ret = new List<ExpressionBlock>();
            ret.AddRange(this.ValueExpressions);
            return ret;
        }
        public override string ToString()
        {
            return "list";
        }
        public override string AsExpString(IFsDataProvider provider)
        {
            var sb = new StringBuilder();
            sb.Append("[");
            
            foreach (var val in this.ValueExpressions)
            {
                sb.Append($"{val.AsExpString(provider)},");
            }
            sb.Append("]");
            return sb.ToString();
        }
    }
}
